### PR TITLE
fix: migrate from decommissioned ipfs gateway to new one

### LINKED ISSUE #11692: Remove decommissioned IPFS gateway from wallet state
### What is this about?

The IPFS gateway cloudflare-ipfs.com was decommissioned on August 14 2024. Prior to #11401, this IPFS gateway was the default for MetaMask mobile users. That was addressed in #11401, but existing users might still have this decommissioned IPFS gateway configured.

We should write a migration to replace the decommissioned gateway with the new default gateway.

The new default gateway shall be 'dweb.link'.

### Scenario

_No response_

### Design

_No response_

### Technical Details

We'll need a state migration here. It should specifically target users who have the decommissioned IPFS gateway selected.

### Threat Modeling Framework

_No response_

### Acceptance Criteria

- The decommissioned IPFS gateway is no longer active for existing users
- The decommissioned IPFS gateway has been replaced by our new default IPFS gateway

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

- Related issue: https://github.com/MetaMask/mobile-planning/issues/1961
- Context: https://github.com/MetaMask/metamask-mobile/pull/11401#discussion_r1775795856

### PR DESCRIPTION
**Pull request by NicolasMassart**
## **Description**

- adds state migration for users with decommissioned IPFS gateway
- adds migration tests
- adds new target gateway (dweb.link) to list and make it replace decommissioned gateway.

>[!NOTE]
> this is only a state migration, the user will not have any visible warnings. It will only show the new gateway as selected instead of the decommissioned one if it was in their state.

## **Related issues**

Fixes #11692

## **Manual testing steps**

```gherkin
Feature: Migrate from decommissioned gateway to dweb.link 
  Scenario: Verify non migrated gateway
    Given an installed app that was built before the decommission of https://cloudflare-ipfs.com/ipfs/ (pre ac7978596566f148fa6892940fb5678170ba0351, for latest is 7.32.0)
    When go to security & privacy settings
    And select https://cloudflare-ipfs.com/ipfs/
    Then see "Your current IPFS gateway is down" as decommissioned gateway does not respond

  Scenario: Check migration
    Given the current PR app installed
    When go to security & privacy settings
    Then see "https://dweb.link/ipfs/" gateway selected
```

### **Before**

Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 22 00 21

### **After**

Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 22 15 31

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/store/migrations/056.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { captureException } from '@sentry/react-native';
+import { isObject } from '@metamask/utils';
+import { ensureValidState } from './util';
+
+export default function migrate(state: unknown) {
+  if (!ensureValidState(state, 56)) {
+    // Increment the migration number as appropriate
```
Comment (tommasini) [inline app/store/migrations/056.ts:7]: NIT: I think we do not need this comment

File: app/store/migrations/056.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    return state;
+  }
+
+  const decommisionedIpfsGateway = 'https://cloudflare-ipfs.com/ipfs/';
+  const newDefaultIpfsGateway = 'https://dweb.link/ipfs/';
+
+  if (decommisionedIpfsGateway === preferencesController.ipfsGateway) {
+    preferencesController.ipfsGateway = newDefaultIpfsGateway;
+  }
+  // Return the modified state
```
Comment (tommasini) [inline app/store/migrations/056.ts:29]: neither this

File: app/store/migrations/056.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        `FATAL ERROR: Migration 56: Invalid PreferencesController state error: '${typeof preferencesController}'`,
+      ),
+    );
+    return state;
+  }
+
+  const decommisionedIpfsGateway = 'https://cloudflare-ipfs.com/ipfs/';
+  const newDefaultIpfsGateway = 'https://dweb.link/ipfs/';
+
+  if (decommisionedIpfsGateway === preferencesController.ipfsGateway) {
```
Comment (tommasini) [inline app/store/migrations/056.ts:26]: I think it's good if we add a conditional operator
```suggestion
  if (decommisionedIpfsGateway === preferencesController?.ipfsGateway) {
```

**Review by tommasini — DISMISSED** [review]
LGTM just left some comments

File: app/store/migrations/056.ts (hunk 1 above)
Comment (NicolasMassart) [reply app/store/migrations/056.ts:7]: I just copied it from previous migration, looks useful reminder and not doing any arm...

File: app/store/migrations/056.ts (hunk 2 above)
Comment (NicolasMassart) [reply app/store/migrations/056.ts:29]: Same as above

**Review by tommasini — APPROVED** [review]

**Pull request merged**